lib/Analysis/Utility.cpp

```diff
@@ -653,61 +678,144 @@ static triton::MakeTensorPtrOp getMakeTensorPtrOpImpl(Operation *op, Value v) {
+                                                   Value v) {
   if (Operation *definingOp = v.getDefiningOp()) {
-    return getMakeTensorPtrOpImpl(definingOp, v);
+    return getMakeTensorPtrOpImpl(helper, definingOp, v);
   } else if (BlockArgument arg = v.cast<BlockArgument>()) {
     unsigned argNum = arg.getArgNumber();
     Operation *argOwner = arg.getOwner()->getParentOp();
 
     if (auto forOp = dyn_cast<scf::ForOp>(argOwner)) {
-      return getMakeTensorPtrOp(
-          forOp.getOperand(argNum + forOp.getNumControlOperands() - 1));
+      return getMakeTensorPtrOp_(
+          helper, forOp.getOperand(argNum + forOp.getNumControlOperands() - 1));
     } else if (auto funcOp = dyn_cast<mlir::triton::FuncOp>(argOwner)) {
       Block *block = arg.getOwner();
       Operation *op;
       int tOrF;
-      std::tie(op, tOrF) = blockToCFOps[block][0];
+      std::tie(op, tOrF) = helper.blockToCFOps[block][0];
       if (auto br = dyn_cast<cf::BranchOp>(op)) {
-        return getMakeTensorPtrOp(br.getDestOperands()[argNum]);
+        return getMakeTensorPtrOp_(helper, br.getDestOperands()[argNum]);
       }
       if (auto condBr = dyn_cast<cf::CondBranchOp>(op)) {
         if (tOrF) {
-          return getMakeTensorPtrOp(condBr.getTrueDestOperands()[argNum]);
+          return getMakeTensorPtrOp_(helper,
+                                     condBr.getTrueDestOperands()[argNum]);
         } else {
-          return getMakeTensorPtrOp(condBr.getFalseDestOperands()[argNum]);
+          return getMakeTensorPtrOp_(helper,
+                                     condBr.getFalseDestOperands()[argNum]);
         }
       }
     } else {
-      return getMakeTensorPtrOp(argOwner->getOperand(argNum));
+      return getMakeTensorPtrOp_(helper, argOwner->getOperand(argNum));
     }
   }
 
   llvm_unreachable("Unable to getMakeTensorPtr()");
 }
 
+triton::MakeTensorPtrOp getMakeTensorPtrOp(Value v) {
+  BlockToCFOpsHelper helper(v);
+  return getMakeTensorPtrOp_(helper, v);
+}
+
+static Value getParentValueWithSameEncoding_(BlockToCFOpsHelper &helper,
+                                             Attribute layout, Value value) {
+  if (Operation *defOp = value.getDefiningOp()) {
+    for (Value operand : defOp->getOperands()) {
+      if (auto type = operand.getType().cast<RankedTensorType>()) {
+        if (type.getEncoding() == layout)
+          return operand;
+      }
+    }
+    // [benzh] some op like make_range has no parent tensor operand.
+    auto type = value.getType().cast<RankedTensorType>();
+    auto sliceLayout =
+        type.getEncoding().cast<triton::gpu::SliceEncodingAttr>();
+    auto shape = sliceLayout.paddedShape(type.getShape());
+    auto eltType = type.getElementType();
+    auto parentType = RankedTensorType::get(shape, eltType, layout);
+    auto context = value.getContext();
+    OpBuilder builder(context);
+    Value dummy = builder.create<arith::ConstantOp>(
```

Comment:
This will insert the operation outside of the module. This doesn't sound right. I'm not able to understand why we need this kind of value and this doesn't seem to be a parent the given value. Could you explain a bit why we need this?